File: majorProject/src/main/java/com/aocc/framework/implementation/AndroidGraphics.java
package com.aocc.framework.implementation;

import java.io.IOException;
import java.io.InputStream;

import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.graphics.BitmapFactory;
import android.graphics.BitmapFactory.Options;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.graphics.Rect;
import android.graphics.RectF;

import com.aocc.framework.GameConstants;
import com.aocc.framework.Graphics;
import com.aocc.framework.Image;
import com.aocc.framework.Viewport;
import com.aocc.majorproject.AssetScale;

//EXCEPT WHERE NOTED, THE FOLLOWING CODE IS SOURCED FROM THE KILOBOLT ANDROID FRAMEWORK

/**
 * Renders the game into an off-screen buffer sized to the letterboxed viewport,
 * then blits once to the surface. World coordinates stay 1280×720; the buffer
 * scale matches {@link Viewport#getScale()} for sharp output without per-primitive
 * surface scaling cost.
 */
public class AndroidGraphics implements Graphics {
    private final AssetManager assets;
    private Canvas canvas;
    private final Paint paint;
    private final Rect srcRect = new Rect();
    private final Rect dstRect = new Rect();
    private final Paint filteredBitmapPaint =
            new Paint(Paint.FILTER_BITMAP_FLAG | Paint.ANTI_ALIAS_FLAG);
    private final Paint strokePaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    private Bitmap frameBuffer;
    private Canvas frameCanvas;
    private float frameScale = 1f;

    public AndroidGraphics(AssetManager assets) {
        this.assets = assets;
        this.paint = new Paint();
    }

    public void ensureFramebuffer(Viewport viewport) {
        Rect dest = viewport.getLetterboxDestRect();
        int width = dest.width();
        int height = dest.height();
        if (width <= 0 || height <= 0) {
            return;
        }
        if (frameBuffer == null
                || frameBuffer.getWidth() != width
                || frameBuffer.getHeight() != height) {
            if (frameBuffer != null) {
                frameBuffer.recycle();
            }
            frameBuffer = Bitmap.createBitmap(width, height, Config.RGB_565);
            frameCanvas = new Canvas(frameBuffer);
        }
        frameScale = viewport.getScale();
    }

    public Bitmap getFrameBuffer() {
        return frameBuffer;
    }

    /** Begin painting into the off-screen buffer for this frame. */
    public void beginFrame(Viewport viewport) {
        ensureFramebuffer(viewport);
        if (frameCanvas == null) {
            throw new IllegalStateException("Framebuffer not ready");
        }
        canvas = frameCanvas;
        canvas.drawColor(Color.BLACK);
        canvas.save();
        canvas.scale(frameScale, frameScale);
    }

    public void endFrame() {
        if (canvas != null) {
            canvas.restore();
            canvas = null;
        }
    }

    private Canvas requireCanvas() {
        if (canvas == null) {
            throw new IllegalStateException("beginFrame() must be called before drawing");
        }
        return canvas;
    }

    @Override
    public Image newImage(String fileName, ImageFormat format) {
        return newImage(fileName, format, 1);
    }

    @Override
    public Image newImage(String fileName, ImageFormat format, int pixelScale) {
        boolean has2x = assetExists(AssetScale.TWO_X_FOLDER + fileName);
        String path = AssetScale.resolvePath(fileName, pixelScale, has2x);
        int effectiveScale = AssetScale.effectivePixelScale(pixelScale, has2x);
        return decodeImage(path, format, effectiveScale);
    }

    private boolean assetExists(String path) {
        try {
            assets.open(path).close();
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    private Image decodeImage(String fileName, ImageFormat format, int pixelScale) {
        Config config;
        if (format == ImageFormat.RGB565)
            config = Config.RGB_565;
        else if (format == ImageFormat.ARGB4444)
            config = Config.ARGB_4444;
        else
            config = Config.ARGB_8888;

        Options options = new Options();
        options.inPreferredConfig = config;

        InputStream in = null;
        Bitmap bitmap;
        try {
            in = assets.open(fileName);
            bitmap = BitmapFactory.decodeStream(in, null, options);
            if (bitmap == null)
                throw new RuntimeException("Couldn't load bitmap from asset '"
                        + fileName + "'");
        } catch (IOException e) {
            throw new RuntimeException("Couldn't load bitmap from asset '"
                    + fileName + "'");
        } finally {
            if (in != null) {
                try {
                    in.close();
                } catch (IOException e) {
                }
            }
        }

        if (bitmap.getConfig() == Config.RGB_565)
            format = ImageFormat.RGB565;
        else if (bitmap.getConfig() == Config.ARGB_4444)
            format = ImageFormat.ARGB4444;
        else
            format = ImageFormat.ARGB8888;

        return new AndroidImage(bitmap, format, pixelScale);
    }

    @Override
    public void clearScreen(int color) {
        Canvas c = requireCanvas();
        paint.setColor(color);
        paint.setStyle(Style.FILL);
        c.drawRect(0, 0, GameConstants.WORLD_WIDTH, GameConstants.WORLD_HEIGHT, paint);
    }

    @Override
    public void drawLine(int x, int y, int x2, int y2, int color) {
        paint.setColor(color);
        requireCanvas().drawLine(x, y, x2, y2, paint);
    }

    @Override
    public void drawCircle(float x, float y, float radius, int color) {
        paint.setColor(color);
        requireCanvas().drawCircle(x, y, radius, paint);
    }

    @Override
    public void drawArc(RectF oval, float startAngle, float sweepAngle,
            boolean useCenter, int color) {
        paint.setAntiAlias(true);
        paint.setColor(color);
        requireCanvas().drawArc(oval, startAngle, sweepAngle, useCenter, paint);
    }

    @Override
    public void drawButton(int x, int y, int height, int color, String text) {
        requireCanvas().drawText(text, x, y, paint);
    }

    @Override
    public void drawRect(int x, int y, int width, int height, int color) {
        paint.setColor(color);
        paint.setStyle(Style.FILL);
        paint.setAntiAlias(true);
        requireCanvas().drawRect(x, y, x + width - 1, y + height - 1, paint);
    }

    @Override
    public void drawRectOutline(int x, int y, int width, int height, int color,
            float strokeWidth) {
        strokePaint.setColor(color);
        strokePaint.setStyle(Style.STROKE);
        strokePaint.setStrokeWidth(strokeWidth);
        float half = strokeWidth / 2f;
        requireCanvas().drawRect(
                x + half, y + half,
                x + width - half, y + height - half,
                strokePaint);
    }

    @Override
    public void drawCircleOutline(float centerX, float centerY, float radius, int color,
            float strokeWidth) {
        strokePaint.setColor(color);
        strokePaint.setStyle(Style.STROKE);
        strokePaint.setStrokeWidth(strokeWidth);
        requireCanvas().drawCircle(centerX, centerY, radius - strokeWidth / 2f, strokePaint);
    }

    @Override
    public void drawARGB(int a, int r, int g, int b) {
        paint.setStyle(Style.FILL);
        paint.setARGB(a, r, g, b);
        requireCanvas().drawRect(0, 0, GameConstants.WORLD_WIDTH, GameConstants.WORLD_HEIGHT, paint);
    }

    @Override
    public void drawString(String text, int x, int y, int color, Paint textPaint) {
        textPaint.setColor(color);
        requireCanvas().drawText(text, x, y, textPaint);
    }

    public void drawImage(Image image, int x, int y, int srcX, int srcY,
            int srcWidth, int srcHeight) {
        AndroidImage androidImage = (AndroidImage) image;
        int pixelScale = androidImage.getPixelScale();

        srcRect.left = srcX;
        srcRect.top = srcY;
        srcRect.right = srcX + srcWidth;
        srcRect.bottom = srcY + srcHeight;

        if (pixelScale == 1 && srcX == 0 && srcY == 0
                && srcWidth == androidImage.getBitmap().getWidth()
                && srcHeight == androidImage.getBitmap().getHeight()) {
            requireCanvas().drawBitmap(androidImage.getBitmap(), x, y, null);
            return;
        }

        dstRect.left = x;
        dstRect.top = y;
        dstRect.right = x + srcWidth / pixelScale;
        dstRect.bottom = y + srcHeight / pixelScale;

        requireCanvas().drawBitmap(androidImage.getBitmap(), srcRect, dstRect,
                filteredBitmapPaint);
    }

    @Override
    public void drawImage(Image image, int x, int y) {
        AndroidImage androidImage = (AndroidImage) image;
        if (androidImage.getPixelScale() == 1) {
            requireCanvas().drawBitmap(androidImage.getBitmap(), x, y, null);
            return;
        }
        dstRect.left = x;
        dstRect.top = y;
        dstRect.right = x + androidImage.getWidth();
        dstRect.bottom = y + androidImage.getHeight();
        requireCanvas().drawBitmap(androidImage.getBitmap(), null, dstRect, filteredBitmapPaint);
    }

    public void drawScaledImage(Image image, int x, int y, int width, int height, int srcX, int srcY, int srcWidth, int srcHeight) {
        AndroidImage androidImage = (AndroidImage) image;

        srcRect.left = srcX;
        srcRect.top = srcY;
        srcRect.right = srcX + srcWidth;
        srcRect.bottom = srcY + srcHeight;

        dstRect.left = x;
        dstRect.top = y;
        dstRect.right = x + width;
        dstRect.bottom = y + height;

        requireCanvas().drawBitmap(androidImage.getBitmap(), srcRect, dstRect, filteredBitmapPaint);
    }

    @Override
    public int getWidth() {
        return GameConstants.WORLD_WIDTH;
    }

    @Override
    public int getHeight() {
        return GameConstants.WORLD_HEIGHT;
    }
}

File: majorProject/src/main/kotlin/com/aocc/framework/implementation/AndroidGraphics.kt
package com.aocc.framework.implementation

import android.content.res.AssetManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.RectF
import com.aocc.framework.GameConstants
import com.aocc.framework.Graphics
import com.aocc.framework.Graphics.ImageFormat
import com.aocc.framework.Image
import com.aocc.framework.Viewport
import com.aocc.majorproject.AssetScale
import java.io.IOException
import java.io.InputStream

//EXCEPT WHERE NOTED, THE FOLLOWING CODE IS SOURCED FROM THE KILOBOLT ANDROID FRAMEWORK

/**
 * Renders the game into an off-screen buffer sized to the letterboxed viewport,
 * then blits once to the surface. World coordinates stay 1280×720; the buffer
 * scale matches [Viewport.getScale] for sharp output without per-primitive
 * surface scaling cost.
 */
class AndroidGraphics(private val assets: AssetManager) : Graphics {

    private var canvas: Canvas? = null
    private val paint = Paint()
    private val srcRect = Rect()
    private val dstRect = Rect()
    private val filteredBitmapPaint =
        Paint(Paint.FILTER_BITMAP_FLAG or Paint.ANTI_ALIAS_FLAG)
    private val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG)

    var frameBuffer: Bitmap? = null
        private set
    private var frameCanvas: Canvas? = null
    private var frameScale: Float = 1f

    fun ensureFramebuffer(viewport: Viewport) {
        val dest = viewport.getLetterboxDestRect()
        val width = dest.width()
        val height = dest.height()
        if (width <= 0 || height <= 0) {
            return
        }
        val current = frameBuffer
        if (current == null || current.width != width || current.height != height) {
            current?.recycle()
            val fresh = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
            frameBuffer = fresh
            frameCanvas = Canvas(fresh)
        }
        frameScale = viewport.getScale()
    }

    /** Begin painting into the off-screen buffer for this frame. Returns false if not ready yet. */
    fun beginFrame(viewport: Viewport): Boolean {
        ensureFramebuffer(viewport)
        val frameCanvas = frameCanvas ?: return false
        canvas = frameCanvas
        frameCanvas.drawColor(Color.BLACK)
        frameCanvas.save()
        frameCanvas.scale(frameScale, frameScale)
        return true
    }

    fun endFrame() {
        canvas?.let {
            it.restore()
            canvas = null
        }
    }

    private fun requireCanvas(): Canvas {
        return canvas ?: throw IllegalStateException("beginFrame() must be called before drawing")
    }

    override fun newImage(fileName: String, format: ImageFormat): Image {
        return newImage(fileName, format, 1)
    }

    override fun newImage(fileName: String, format: ImageFormat, pixelScale: Int): Image {
        val has2x = assetExists(AssetScale.TWO_X_FOLDER + fileName)
        val path = AssetScale.resolvePath(fileName, pixelScale, has2x)
        val effectiveScale = AssetScale.effectivePixelScale(pixelScale, has2x)
        return decodeImage(path, format, effectiveScale)
    }

    private fun assetExists(path: String): Boolean {
        return try {
            assets.open(path).close()
            true
        } catch (e: IOException) {
            false
        }
    }

    private fun decodeImage(fileName: String, format: ImageFormat, pixelScale: Int): Image {
        var resolvedFormat = format
        val config = when (format) {
            ImageFormat.RGB565 -> Bitmap.Config.RGB_565
            ImageFormat.ARGB4444 -> Bitmap.Config.ARGB_4444
            else -> Bitmap.Config.ARGB_8888
        }

        val options = BitmapFactory.Options()
        options.inPreferredConfig = config

        var `in`: InputStream? = null
        val bitmap: Bitmap
        try {
            `in` = assets.open(fileName)
            bitmap = BitmapFactory.decodeStream(`in`, null, options)
                ?: throw RuntimeException("Couldn't load bitmap from asset '$fileName'")
        } catch (e: IOException) {
            throw RuntimeException("Couldn't load bitmap from asset '$fileName'")
        } finally {
            try {
                `in`?.close()
            } catch (e: IOException) {
            }
        }

        resolvedFormat = when (bitmap.config) {
            Bitmap.Config.RGB_565 -> ImageFormat.RGB565
            Bitmap.Config.ARGB_4444 -> ImageFormat.ARGB4444
            else -> ImageFormat.ARGB8888
        }

        return AndroidImage(bitmap, resolvedFormat, pixelScale)
    }

    override fun clearScreen(color: Int) {
        val c = requireCanvas()
        paint.color = color
        paint.style = Paint.Style.FILL
        c.drawRect(0f, 0f, GameConstants.WORLD_WIDTH.toFloat(), GameConstants.WORLD_HEIGHT.toFloat(), paint)
    }

    override fun drawLine(x: Int, y: Int, x2: Int, y2: Int, color: Int) {
        paint.color = color
        requireCanvas().drawLine(x.toFloat(), y.toFloat(), x2.toFloat(), y2.toFloat(), paint)
    }

    override fun drawCircle(x: Float, y: Float, radius: Float, color: Int) {
        paint.color = color
        requireCanvas().drawCircle(x, y, radius, paint)
    }

    override fun drawArc(
        oval: RectF,
        startAngle: Float,
        sweepAngle: Float,
        useCenter: Boolean,
        color: Int
    ) {
        paint.isAntiAlias = true
        paint.color = color
        requireCanvas().drawArc(oval, startAngle, sweepAngle, useCenter, paint)
    }

    override fun drawButton(x: Int, y: Int, height: Int, color: Int, text: String) {
        requireCanvas().drawText(text, x.toFloat(), y.toFloat(), paint)
    }

    override fun drawRect(x: Int, y: Int, width: Int, height: Int, color: Int) {
        paint.color = color
        paint.style = Paint.Style.FILL
        paint.isAntiAlias = true
        requireCanvas().drawRect(
            x.toFloat(),
            y.toFloat(),
            (x + width - 1).toFloat(),
            (y + height - 1).toFloat(),
            paint
        )
    }

    override fun drawRectOutline(
        x: Int,
        y: Int,
        width: Int,
        height: Int,
        color: Int,
        strokeWidth: Float
    ) {
        strokePaint.color = color
        strokePaint.style = Paint.Style.STROKE
        strokePaint.strokeWidth = strokeWidth
        val half = strokeWidth / 2f
        requireCanvas().drawRect(
            x + half,
            y + half,
            x + width - half,
            y + height - half,
            strokePaint
        )
    }

    override fun drawCircleOutline(
        centerX: Float,
        centerY: Float,
        radius: Float,
        color: Int,
        strokeWidth: Float
    ) {
        strokePaint.color = color
        strokePaint.style = Paint.Style.STROKE
        strokePaint.strokeWidth = strokeWidth
        requireCanvas().drawCircle(centerX, centerY, radius - strokeWidth / 2f, strokePaint)
    }

    override fun drawARGB(a: Int, r: Int, g: Int, b: Int) {
        paint.style = Paint.Style.FILL
        paint.setARGB(a, r, g, b)
        requireCanvas().drawRect(
            0f,
            0f,
            GameConstants.WORLD_WIDTH.toFloat(),
            GameConstants.WORLD_HEIGHT.toFloat(),
            paint
        )
    }

    override fun drawString(text: String, x: Int, y: Int, color: Int, textPaint: Paint) {
        textPaint.color = color
        requireCanvas().drawText(text, x.toFloat(), y.toFloat(), textPaint)
    }

    override fun drawImage(
        image: Image,
        x: Int,
        y: Int,
        srcX: Int,
        srcY: Int,
        srcWidth: Int,
        srcHeight: Int
    ) {
        val androidImage = image as AndroidImage
        val pixelScale = androidImage.pixelScale

        srcRect.left = srcX
        srcRect.top = srcY
        srcRect.right = srcX + srcWidth
        srcRect.bottom = srcY + srcHeight

        if (pixelScale == 1 && srcX == 0 && srcY == 0
            && srcWidth == androidImage.bitmap.width
            && srcHeight == androidImage.bitmap.height
        ) {
            requireCanvas().drawBitmap(androidImage.bitmap, x.toFloat(), y.toFloat(), null)
            return
        }

        dstRect.left = x
        dstRect.top = y
        dstRect.right = x + srcWidth / pixelScale
        dstRect.bottom = y + srcHeight / pixelScale

        requireCanvas().drawBitmap(androidImage.bitmap, srcRect, dstRect, filteredBitmapPaint)
    }

    override fun drawImage(image: Image, x: Int, y: Int) {
        val androidImage = image as AndroidImage
        if (androidImage.pixelScale == 1) {
            requireCanvas().drawBitmap(androidImage.bitmap, x.toFloat(), y.toFloat(), null)
            return
        }
        dstRect.left = x
        dstRect.top = y
        dstRect.right = x + androidImage.width
        dstRect.bottom = y + androidImage.height
        requireCanvas().drawBitmap(androidImage.bitmap, null, dstRect, filteredBitmapPaint)
    }

    fun drawScaledImage(
        image: Image,
        x: Int,
        y: Int,
        width: Int,
        height: Int,
        srcX: Int,
        srcY: Int,
        srcWidth: Int,
        srcHeight: Int
    ) {
        val androidImage = image as AndroidImage

        srcRect.left = srcX
        srcRect.top = srcY
        srcRect.right = srcX + srcWidth
        srcRect.bottom = srcY + srcHeight

        dstRect.left = x
        dstRect.top = y
        dstRect.right = x + width
        dstRect.bottom = y + height

        requireCanvas().drawBitmap(androidImage.bitmap, srcRect, dstRect, filteredBitmapPaint)
    }

    override val width: Int
        get() = GameConstants.WORLD_WIDTH

    override val height: Int
        get() = GameConstants.WORLD_HEIGHT
}
